mcp.go

```diff
@@ -0,0 +1,909 @@
+				property["type"] = "array"
+				property["items"] = map[string]interface{}{"type": "string"}
+			} else {
+				property["type"] = "object"
+				property["additionalProperties"] = map[string]interface{}{
+					"type": "string", // Default to string value type for maps
+				}
+				if valueType == "stringToInt" || valueType == "stringToInt64" {
+					property["additionalProperties"] = map[string]interface{}{
+						"type": "integer",
+					}
+				}
+			}
+		// Handle custom serpent types
+		default:
+			// Check for known serpent custom types (Enum, EnumArray)
+			if enum, ok := opt.Value.(*Enum); ok {
+				property["type"] = "string"
+				property["enum"] = enum.Choices
+			} else if enumArray, ok := opt.Value.(*EnumArray); ok {
+				property["type"] = "array"
+				property["items"] = map[string]interface{}{
+					"type": "string",
+					"enum": enumArray.Choices,
+				}
+			} else {
+				// Fallback for unknown types
+				property["type"] = "string"
+				fmt.Fprintf(s.stderr, "Warning: Unknown pflag type '%s' for option '%s', defaulting to string\n", valueType, opt.Flag)
+			}
+		}
+
+		// Add the property definition
+		properties[opt.Flag] = property
+
+		// Add to required list if Required is true AND Default is not set
+		// (as per comment in option.go)
+		if opt.Required && opt.Default == "" {
+			requiredList = append(requiredList, opt.Flag)
+		}
+	}
+
+	// Update required field only if it's not empty
+	if len(requiredList) > 0 {
+		schema["required"] = requiredList
+	} else {
+		// Remove the empty required array if no options are required
+		delete(schema, "required")
+	}
+
+	return json.MarshalIndent(schema, "", "  ") // Use MarshalIndent for readability
+}
+
+// sendErrorResponse sends an error JSON-RPC response
+func (s *MCPServer) sendErrorResponse(id interface{}, code int, message string, data interface{}) {
+	var dataBytes json.RawMessage
+	if data != nil {
+		var err error
+		dataBytes, err = json.Marshal(data)
+		if err != nil {
+			// If we can't marshal the data, just ignore it
+			dataBytes = nil
+		}
+	}
+
+	response := JSONRPC2Response{
+		JSONRPC: "2.0",
+		ID:      id,
+		Error: &JSONRPC2Error{
+			Code:    code,
+			Message: message,
+			Data:    dataBytes,
+		},
+	}
+
+	s.sendResponse(response)
+}
+
+// sendResponse sends a JSON-RPC response to stdout
+func (s *MCPServer) sendResponse(response JSONRPC2Response) {
+	responseBytes, err := json.Marshal(response)
+	if err != nil {
+		fmt.Fprintf(s.stderr, "Failed to marshal response: %v\n", err)
+		return
+	}
+
+	fmt.Fprintln(s.stdout, string(responseBytes))
+}
+
+// unmarshalParamsLenient unmarshals JSON parameters in a lenient way, handling omitted optional parameters.
+// If the raw JSON is empty, null, or "[]", it will use the default zero value for the target struct.
+// If the JSON contains invalid data (not empty), the original error is returned.
+func (s *MCPServer) unmarshalParamsLenient(data json.RawMessage, v interface{}) error {
```

Comment:
I don't think this needs to be a struct method, which means we can make it `func UnmarshalParamsLenient[T any](data json.RawMessage) (T, error)`